scipy/linalg/decomp_svd.py

```diff
@@ -264,35 +264,76 @@ def diagsvd(s, M, N):
 
 # Orthonormal decomposition
 
-def orth(A):
+def orth(A, rcond=None):
     """
     Construct an orthonormal basis for the range of A using SVD
 
     Parameters
     ----------
     A : (M, N) array_like
         Input array
+    rcond : float, optional
+        Relative condition number. Singular values ``s`` smaller than
+        ``rcond * max(M,N) * s[0]`` are considered zero.
+        Default: floating point eps.
 
     Returns
     -------
     Q : (M, K) ndarray
         Orthonormal basis for the range of A.
-        K = effective rank of A, as determined by automatic cutoff
+        K = effective rank of A, as determined by rcond
 
     See also
     --------
     svd : Singular value decomposition of a matrix
+    null_space : Matrix null space
 
     """
     u, s, vh = svd(A, full_matrices=False)
     M, N = A.shape
-    eps = numpy.finfo(float).eps
-    tol = max(M, N) * numpy.amax(s) * eps
+    if rcond is None:
+        rcond = numpy.finfo(float).eps
+    tol = max(M, N) * numpy.amax(s) * rcond
     num = numpy.sum(s > tol, dtype=int)
     Q = u[:, :num]
     return Q
 
 
+def null_space(A, rcond=None):
+    """
+    Construct an orthonormal basis for the null space of A using SVD
+
+    Parameters
+    ----------
+    A : (M, N) array_like
+        Input array
+    rcond : float, optional
+        Relative condition number. Singular values ``s`` smaller than
+        ``rcond * max(M,N) * s[0]`` are considered zero.
+        Default: floating point eps.
+
+    Returns
+    -------
+    Q : (M, K) ndarray
+        Orthonormal basis for the null space of A.
+        K = effective rank of A, as determined by rcond
+
+    See also
+    --------
+    svd : Singular value decomposition of a matrix
+    orth : Matrix range
+
+    """
+    u, s, vh = svd(A, full_matrices=True)
+    M, N = A.shape
+    if rcond is None:
+        rcond = numpy.finfo(float).eps
+    tol = max(M, N) * numpy.amax(s) * rcond
```

Comment:
why is this multiplied by max(M, N)?
Is this usual for threshold in svd with rcond?

Reply:
I was matching the threshold in `orth`, and it's also used in lstsq et al., see https://github.com/numpy/numpy/pull/9582/files

Reply:
hmm, maybe it indeed should only be in the default value.

Reply:
thanks for the numpy link
I'm not sure I like that change, This will affect many results in statsmodels, where we usually use numpy defaults, (although we use more often pinv than lstsq)
Especially when M is a few million.

Reply:
@josef-pkt These bounds are kind of standard. (well standard in the sense "that's what matlab does"). NumPy also picked these up for [matrix_rank tolerance](https://github.com/numpy/numpy/blob/master/numpy/linalg/linalg.py#L1536) which should be inline with null basis column number. You can see more about it here https://www.ma.utexas.edu/documentation/lapack/node92.html

Reply:
The lapack links says "p(m , n) is a modestly growing function of m and n"
My question is whether linear growth in m is not too much (e.g. sqrt(m) or log(M)?). statsmodels has some issues that constant in m (rcond independent of m) is too small.

Reply:
@josef-pkt I'm really not experienced in dense matrices of those sizes or knowledgeable whether this relationship holds up to millions. But `max(.,.)` is pretty ubiquitous in linalg libraries.

Reply:
For consistency with orth and other routines at the minimum, I think we should use the max(M,N) for the default cutoff here.

Reply:
I agree that consistency in scipy/numpy is important.
(My problem is whether we keep the defaults or should override them consistently in statsmodels, now that they will be changed by numpy anyway, and whether I'm even able to find a more appropriate solution for us. A few million observations doesn't seem to put much strain on linear algebra solutions anymore even on desktops.)

Reply:
Just to bookmark this numpy.linalg.cond is also buggy. FYI: https://github.com/numpy/numpy/issues/8090